packages/react-mongo/react-mongo.ts

```diff
@@ -0,0 +1,63 @@
+import { Meteor } from 'meteor/meteor'
+import { Mongo } from 'meteor/mongo'
+import { Tracker } from 'meteor/tracker'
+import { useEffect, useMemo, useReducer, useRef, DependencyList } from 'react'
+
+const fur = (x: number): number => x + 1
+const useForceUpdate = () => useReducer(fur, 0)[1]
+
+const useSubscriptionClient = (factory: () => Meteor.SubscriptionHandle, deps: DependencyList = []) => {
+  const forceUpdate = useForceUpdate()
+  const subscription = useRef<Meteor.SubscriptionHandle>({
+    stop() {},
+    ready: () => false
+  })
+
+  useEffect(() => {
+    subscription.current = factory()
+    const computation = Tracker.autorun(() => {
+      if (subscription.current.ready()) forceUpdate()
+    })
+
+    return () => {
+      computation.stop()
```

Comment:
I think we're missing a `subscription.current.stop()` here.

Reply:
It's letting the computation stop it intentionally. Meteor does a bunch of stuff internally to recycle subscriptions, and I wonder if you are running in to that? I'm not super familiar with how some of that magic works, but I think it's important to retain it. Still, we can look in to how that works, and make a more informed decision about whether to intentionally stop the subscription.

I'm curious if the subscription would be stopped when using `useTracker`.

Reply:
Oh cool, I didn't know that. I just tested with the `useTracker` alternative:

```ts
useTracker( () => {
  Meteor.subscribe( 'board.projects', boardId )
}, [ boardId ] )
```

And this stopped the subscriptions immediately when switching. I must say that although I don't know about any performance implications with subscription recycling, it feels a lot more predictable if the subscription is just stopped.

Reply:
Looking at [the `Tracker` code](https://github.com/meteor/meteor/blob/devel/packages/tracker/tracker.js) (which is 6 years old by the way, haha), I don't think it even knows about subscriptions, let alone stops/recycles them. I think stopping the computation only stops it looking for updates in the subscription. Could be wrong here though of course.

Reply:
I've read through that a few times, but I'm unsure if I ever found a place where Subscriptions are stopped. I do know that `withTracker` never stopped subscriptions, and neither does `useTracker`. I'm really not sure how that works, but I assume it's intentional.

The question is really, what do we do about it now? I could easily stop the subscription whenever deps change, but what's the effect of that? Does it dump all documents immediately, then run the next subscription? What if 80% of the documents overlap - would it then have to load everything all over again?

There are some questions to answer before implementing that type of change. But, it may also be possible to simply work around the problem, by adding an ability to stop the subscription yourself, or adding a config setting.

Whatever the case, I want to know more about how this all works on the Meteor side.

Reply:
Fair enough! I'll do some more testing to try to help.

I happen to have two boards with an overlapping document. Board 1 has two Projects, Board 2 has 4 Projects of which 1 is also on Board 1. When doing a manual `subscription.stop()`, all documents are indeed dumped and reloaded (bottom to top, note project ID `FcSFkRwNHtWpcDtSe`):

<img width="653" alt="image" src="https://user-images.githubusercontent.com/159500/97841284-c7cb2580-1ce5-11eb-88c4-c5ad9e9f74b9.png">


And you're absolutely right, with `useTracker` there definitely seems to be some more smartness involved:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/159500/97840997-4ffcfb00-1ce5-11eb-8fb9-bdffc0948d9f.png">

Note that the overlapping document is reused, and the one from Board 1 that doesn't overlap is only dropped at the very end of loading the Board 2 documents. Even though the subscription is closed as good as immediately, it's not kept open on the client or anything.

Looking into it further, I think this is the crucial difference that allows the server to recycle these documents:

`useTracker` | `useSubscription` + `subscription.stop()`
--------|----------
<img width="659" alt="image" src="https://user-images.githubusercontent.com/159500/97841788-d108c200-1ce6-11eb-9a18-5027357f89aa.png">|<img width="656" alt="image" src="https://user-images.githubusercontent.com/159500/97841994-3197ff00-1ce7-11eb-8a9f-3390f820bae6.png">
New subscription (`DAoL`) is started, then old subscription (`bpGr`) is stopped|Old is stopped first, then new is started

Reply:
Allow me to follow up.. I feel a bit ashamed because this might have been a false alarm 😬

Looking at the messages and Minimongo, the subscription _is_ closed properly, at the same time (with recycling of documents) `useTracker` does it. I also see the non-overlapping documents are removed from Minimongo. So this must simply be an issue with the Meteor Devtools not picking up the way these subscriptions are closed. Because even though I see the `nosub` DDP messages, the subscription is still listed on the Subscriptions tab. Probably something to do with the way it is monkey patched.

I'll loop in @leonardoventurini; let me know if you'd like a PR for this and if you have some pointers.

Reply:
> I'm really not sure how that works, but I assume it's intentional.

Funny though, that after 6 years of working with Meteor, `Tracker` is still a black box even to us. 🙃 Something to think about @MattH1611.

Reply:
> this must simply be an issue with the Meteor Devtools not picking up the way these subscriptions are closed

Scratch that (sorry for the spam @leonardoventurini), there really does seem to be an issue with the hook. I disabled my other subscriptions to reduce the noise, and then clicked through some boards:

`useTracker` | `useSubscription`
--------|----------
<img width="648" alt="image" src="https://user-images.githubusercontent.com/159500/97845213-999d1400-1cec-11eb-9404-adb20d52d7ca.png">|<img width="647" alt="image" src="https://user-images.githubusercontent.com/159500/97845081-635f9480-1cec-11eb-99a6-596ff98c781a.png">

Reply:
@CaptainN maybe the problem is that the subscription isn't initialized inside the `Tracker.autorun()` context? When I move that line down:

```ts
useEffect( () => {
  // subscription.current = factory() moved this down 👇
  const computation = Tracker.autorun( () => {
    subscription.current = factory()
    if ( subscription.current.ready() ) forceUpdate()
  })

  return () => {
    computation.stop()
  }
}, deps )
```

Everything seems to work properly, and similar in DDP messages to `useTracker`.

Reply:
Oh snap! That seems obvious now that you said it! Definitely will change.

Reply:
Bit late to the party here, but regarding the Tracker mystery: Tracker doesn't know about subscriptions, but Subscriptions know about Tracker :).

When you call subscribe, Meteor checks whether you're currently in an autorun function. If so, it adds hooks to Tracker that mark the subscription as inactive when the autorun function is ran again. If the same subscription is called (= same name & params), it is marked again as active. After the autorun completes, all subscriptions that are inactive are stopped. It's quite elegant and works fine even when nesting autorun functions.

See: https://github.com/meteor/meteor/blob/devel/packages/ddp-client/common/livedata_connection.js#L385-L392
and: https://github.com/meteor/meteor/blob/devel/packages/ddp-client/common/livedata_connection.js#L473-L492

Reply:
What happens if the same subscription is run in two different places in the app at the same time, with. the. same params, and then one of them has their autorun (computation) ended? Is it smart enough to keep the other one active?

Reply:
That would depend on the definition of 'at the same time'.

In single-threaded javascript this can only occur if you would have an async autorun function. Tracker does not handle async autorun functions (since it is from a pre-async/await period). In fact, on the server it will wrap the function in a `_noYieldsAllowed`, so it can guarantee this does not happen using the fiber mechanisms. But it has no similar functionality in the browser.

What would happen is that Tracker sees one autorun as a nested child of another autorun in this case. The impact on subscriptions would be subject to race conditions and is therefore undefined.

Long story short: you can't use async methods as autorun functions. As long as you stick to that, Meteor will handle duplicate subscriptions in multiple autoruns fine.